ADR/0018-apps-continuous-perf-testing.md

```diff
@@ -0,0 +1,69 @@
+
+The approach to make this pass/fail detection possible proposed here, is to use historical results as a benchmark – not only actual results of the performance test, but also monitoring data about the resource usage, etc.
+
+> Example: Imagine you develop a web e-shop application that uses PostgreSQL data backend. Your perf test is browsing through a goods catalog and measures latency of pages. When you want to decide on pass or fail result of this test, you need to check that metrics like below are aligned to previous results of a same test with same configuration:
+>
+> - Average page load latency.
+> - Backend service CPU and memory usage...
+> - PostgreSQL CPU and memory usage…
+> - Internal PostgreSQL metrics like number of commits during the test or average number of database sessions during the test…
+>
+> If any of these metrics does not align to historical results, we mark this test result as a failure.
+
+And even if some test fails, the application team should be able to review the data and change the test definition, so next time a new test result is being evaluated based on the historical results, this new result will be one of them.
+
+The algorithm that says if the current test passes or fails when compared to historical results can vary. It can be:
+
+- Using a floating average of historical data and setting a safe range around it and making sure the new result is in that range.
+- Some other possibilities might evolve later. Above is just an example used now by some Performance teams.
+
+Goal of this ADR is to propose a way for this kind of testing to be implemented in StoneSoup (feature [STONE-679](https://issues.redhat.com/browse/STONE-679)). Even if it would not deprecate full-blown performance and scale testing, having a release gate with some short and small scale performance test is desirable for many application teams.
+
+### Glossary
+
+- SUT: Subject Under Test
+- Horreum: a results storage service: <https://github.com/Hyperfoil/Horreum>
+- CPT: Continuous Performance Testing
+
+## Decision
+
+Let's use this architecture with a single Horreum instance that would be managed by Red Hat and used by all customers.
+
+![Architecture diagram with Horreum](assets/0018-apps-continuous-perf-testing.svg "Architecture diagram with Horreum")
+
+1. Performance test runs in the Tekton pipeline and generates JSON with test parameters and results.
+2. Pipeline gathers configured monitoring metrics and add them to the JSON.
+3. Pipeline uploads the JSON with all the data to Horreum.
+4. Horreum performs result analysis, looking for changes in configured metrics.
+4. Pipeline gets PASS/FAIL decision from Horreum back to the pipeline, so pipeline can return proper result.
+
```

Comment:
Horreum provides a UI, right?

Can you add a section here describing whether or not users are expected to interact with that UI - or if it will be hidden.

(For example, ArgoCD exists behind our Gitops Service APIs, but we conceal it entirely from the user. They cannot access the Argo UI.)

Reply:
I see! Added.